plutarch-extra/Plutarch/Extra/Map.hs

```diff
@@ -0,0 +1,324 @@
+{-# LANGUAGE QuantifiedConstraints #-}
+
+module Plutarch.Extra.Map (
+  -- * Lookup
+  ptryLookup,
+
+  -- * Comparisons
+  pkeysEqual,
+  pkeysEqualUnsorted,
+
+  -- * Modification
+  pupdate,
+  padjust,
+
+  -- * Folds
+  pfoldMapWithKey,
+  pfoldlWithKey,
+
+  -- * Conversion
+  punsortedMapFromFoldable,
+  psortedMapFromFoldable,
+  pkeys,
+
+  -- * Key-value pair manipulation
+  pkvPairKey,
+  pkvPairValue,
+  pkvPairLt,
+) where
+
+import Data.Foldable (foldl')
+import Plutarch.Api.V1.AssocMap (
+  KeyGuarantees (Sorted, Unsorted),
+  PMap (PMap),
+  pdelete,
+  pempty,
+  pinsert,
+  plookup,
+ )
+import Plutarch.Builtin (ppairDataBuiltin)
+import Plutarch.Extra.Maybe (passertPJust)
+import qualified Plutarch.List as PList
+import Plutarch.Prelude
+
+{- | If a value exists at the specified key, apply the function argument to it;
+ otherwise, do nothing.
+
+ This is necessarily linear in the size of the map performance-wise, as we
+ have to scan the entire map to find the key in the worst case.
+-}
+padjust ::
+  forall (k :: PType) (v :: PType) (s :: S).
+  (PIsData k, PEq k, PIsData v) =>
+  Term s ((v :--> v) :--> k :--> PMap 'Unsorted k v :--> PMap 'Unsorted k v)
+padjust = phoistAcyclic $
+  plam $ \f key kvs ->
+    pmatch kvs $ \(PMap kvs') ->
+      pcon . PMap $ PList.pmap # (go # f # key) # kvs'
+  where
+    go ::
+      forall (s' :: S).
+      Term
+        s'
+        ( (v :--> v)
+            :--> k
+            :--> PBuiltinPair (PAsData k) (PAsData v)
+            :--> PBuiltinPair (PAsData k) (PAsData v)
+        )
+    go = phoistAcyclic $
+      plam $ \f target kv ->
+        pif
+          ((pkvPairKey # kv) #== target)
+          (ppairDataBuiltin # (pfstBuiltin # kv) # pdata (f #$ pkvPairValue # kv))
+          kv
+
+{- | As 'pkeysEqual', but requires only 'PEq' constraints for the keys, and
+ works for 'Unsorted' 'PMap's. This requires a number of equality comparisons
+ between keys proportional to the product of the lengths of both arguments:
+ that is, this function is quadratic.
+-}
+pkeysEqualUnsorted ::
+  forall (k :: PType) (a :: PType) (b :: PType) (s :: S).
+  (PIsData k, PIsData a, PIsData b) =>
+  Term s (PMap 'Unsorted k a :--> PMap 'Unsorted k b :--> PBool)
```

Comment:
When is this more efficient than just sorting before checking?

Reply:
It's more efficient in terms of memories

Reply:
You should note that